pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,310 @@
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/objstore"
+	"github.com/improbable-eng/thanos/pkg/runutil"
+	"github.com/pkg/errors"
+	yaml "gopkg.in/yaml.v2"
+)
+
+type Config struct {
+	EndPoint  string `yaml:"endpoint"`
+	Bucket    string `yaml:"bucket"`
+	SecretKey string `yaml:"secret_key"`
+	SecretId  string `yaml:"secret_id"`
+}
+
+type objectInfo struct {
+	key string
+	err error
+}
+
+type Bucket struct {
+	name   string
+	logger log.Logger
+	client *alioss.Client
+	config Config
+	bucket *alioss.Bucket
+}
+
+func NewTestBucket(t testing.TB, location string) (objstore.Bucket, func(), error) {
+	c := configFromEnv()
+	if err := ValidateForTests(c); err != nil {
+		return nil, nil, err
+	}
+	if c.Bucket != "" && os.Getenv("THANOS_ALLOW_EXISTING_BUCKET_USE") == "" {
+		return nil, nil, errors.New("OSS_BUCKET is defined. Normally this tests will create temporary bucket " +
+			"and delete it after test. Unset OSS_BUCKET env variable to use default logic. If you really want to run " +
+			"tests against provided (NOT USED!) bucket, set THANOS_ALLOW_EXISTING_BUCKET_USE=true. WARNING: That bucket " +
+			"needs to be manually cleared. This means that it is only useful to run one test in a time. This is due " +
+			"to safety (accidentally pointing prod bucket for test) as well as aliyun oss not being fully strong consistent.")
+	}
+
+	return NewTestBucketFromConfig(t, location, c, true)
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	if err := b.bucket.PutObject(name, r); err != nil {
+		return errors.Wrap(err, "upload oss object")
+	}
+	return nil
+}
+
+// Delete removes the object with the given name.
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	if err := b.bucket.DeleteObject(name); err != nil {
+		return errors.Wrap(err, "delete oss object")
+	}
+	return nil
+}
+
+func configFromEnv() Config {
+	c := Config{
+		EndPoint:  os.Getenv("OSS_ENDPOINT"),
+		Bucket:    os.Getenv("OSS_BUCKET"),
+		SecretKey: os.Getenv("OSS_SECRET_KEY"),
+		SecretId:  os.Getenv("OSS_SECRET_ID"),
+	}
+
+	return c
+}
+
+// ValidateForTests checks to see the config options for tests are set.
+func ValidateForTests(conf Config) error {
+	if conf.EndPoint == "" ||
+		conf.SecretId == "" ||
+		conf.SecretKey == "" {
+		return errors.New("insufficient oss test configuration information")
+	}
+	return nil
+}
+
+func parseConfig(conf []byte) (Config, error) {
+	var config Config
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return Config{}, err
+	}
+
+	return config, nil
+}
+
+func NewBucket(logger log.Logger, conf []byte, component string) (*Bucket, error) {
+	config, err := parseConfig(conf)
+	if err != nil {
+		return nil, err
+	}
+	if err := validate(config); err != nil {
+		return nil, err
+	}
+
+	client, err := alioss.New(config.EndPoint, config.SecretId, config.SecretKey)
```

Comment:
One suggestion here. It will be better here if we set user-agent when we create oss client.

Reply:
what is user-agent here used for ? can you explain it  ?

Reply:
It is used to indicate the requests are from thanos to oss like what hadoop does. 

// UserAgent specifies UserAgent. The default is aliyun-sdk-go/1.2.0 (windows/-/amd64;go1.5.2).
//
// userAgent    the user agent string.
//
func UserAgent(userAgent string) ClientOption {
	return func(client *Client) {
		client.Config.UserAgent = userAgent
	}
}

Reply:
I think oss should not be acknowledged  with  anything about what is the data  from , It is just a choice of  storage   , so i will not supply user-agent for client  creating  , but ,   thanks for you suggestion .

Reply:
Oh, you misunderstand my purpose. It is only used to distinguish requests sent by go sdk. With this different user agent, we can do some analysis work against thanos. Yes, it does not affect functionality.

Reply:
Could we ignore user-agent parameter . As you said , It does not affect functionality . But , thank you again.